src/interpreter/vm.rs

```diff
@@ -449,3 +494,345 @@ where
+        msg_receipt: Receipt,
+        gas_cost: TraceMessageGasCost,
+        execution_trace: Option<Trace>,
+        error: String,
+        duration: u64,
+    }
+
+    impl TraceMessageInvocation {
+        pub(super) fn new(msg_cid: Cid, msg: &Message, ret: &ApplyRet) -> Self {
+            let trace = build_exec_trace(ret.exec_trace());
+            Self {
+                msg_cid,
+                msg: msg.clone(),
+                msg_receipt: ret.msg_receipt(),
+                gas_cost: TraceMessageGasCost::new(msg, ret),
+                execution_trace: trace,
+                error: ret.failure_info().unwrap_or_default(),
+                duration: 0,
+            }
+        }
+
+        pub(super) fn from_implicit(msg_cid: Cid, msg: &Message, ret: &ApplyRet) -> Self {
+            Self {
+                gas_cost: TraceMessageGasCost::from_implicit(msg, ret),
+                ..Self::new(msg_cid, msg, ret)
+            }
+        }
+    }
+
+    // We match Lotus structures and code to have similar json trace and an easier diff:
+    // https://github.com/filecoin-project/filecoin-ffi/blob/v1.23.0/rust/src/fvm/machine.rs#L391
+    #[derive(PartialEq, Clone, Debug, Serialize, Deserialize)]
+    #[serde(rename_all = "PascalCase")]
+    struct TraceGasCharge {
+        name: Cow<'static, str>,
+        #[serde(rename = "tg")]
+        total_gas: u64,
+        #[serde(rename = "cg")]
+        compute_gas: u64,
+        #[serde(rename = "sg")]
+        other_gas: u64,
+        #[serde(rename = "tt")]
+        duration_nanos: u64,
+    }
+
+    #[derive(Default, PartialEq, Clone, Debug, Serialize, Deserialize)]
+    #[serde(rename_all = "PascalCase")]
+    struct Trace {
+        msg: TraceMessage,
+        #[serde(rename = "MsgRct")]
+        msg_ret: TraceReturn,
+        gas_charges: Vec<TraceGasCharge>,
+        subcalls: Vec<Trace>,
+    }
+
+    #[derive(Default, PartialEq, Clone, Debug, Serialize, Deserialize)]
+    #[serde(rename_all = "PascalCase")]
+    struct TraceMessage {
+        #[serde(with = "crate::lotus_json")]
+        from: Address,
+        #[serde(with = "crate::lotus_json")]
+        to: Address,
+        #[serde(with = "crate::lotus_json")]
+        value: TokenAmount,
+        #[serde(rename = "Method")]
+        method_num: MethodNum,
+        #[serde(with = "crate::lotus_json")]
+        params: Vec<u8>,
+        params_codec: u64,
+    }
+
+    #[derive(PartialEq, Clone, Debug, Serialize, Deserialize)]
+    #[serde(rename_all = "PascalCase")]
+    struct TraceReturn {
+        exit_code: ExitCode,
+        #[serde(rename = "Return", with = "crate::lotus_json")]
+        return_data: Vec<u8>,
+        return_codec: u64,
+    }
+
+    impl Default for TraceReturn {
+        fn default() -> Self {
+            Self {
+                exit_code: ExitCode::OK,
+                return_data: vec![],
+                return_codec: 0,
+            }
+        }
+    }
+
+    fn build_lotus_trace(
```

Comment:
This function needs to be documented and refactored. What it does is fairly simple: It takes a flat array of `enter`, `return`, and `charge` events. Then, it arranges those events in a tree. For example: `enter X, charge 10, return, enter Y, charge 20, return` becomes `[Enter X [Charge 10], Enter Y [Charge 20]]`.

This function should _not_ use an iterator for `trace_iter`. It can be written as fairly simple and idiomatic Rust.